pghoard/transfer.py

```diff
@@ -39,6 +40,105 @@ class TransferOperation(StrEnum):
+
+
+class OperationEventProgressTracker(PGHoardThread):
+    """
+    Monitors the progress of operation events and updates metrics via clients (telegraph, prometheus, Pushgateway)
+    """
+
+    CHECK_FREQUENCY: int = 5  # check every 5 seconds for progress
+    WARNING_TIMEOUT: int = 5 * 60  # log a warning in case there is no progress during last 5 minutes
+
+    def __init__(
+        self,
+        *,
+        metrics: Metrics,
+        metric_name: str,
+        operation: TransferOperation,
+        file_size: float,
+        tags: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        self.metrics = metrics
+        self.log = logging.getLogger("OperationEventProgressTracker")
+        self.metric_name = metric_name
+        self.tags = tags
+        self.operation = operation
+        self.file_size = file_size
+
+        self.running = True
+        self._increments = []
+        self._started_at: Optional[float] = None
+        super().__init__()
+
+    def stop(self) -> None:
+        self.running = False
+
+    def run_safe(self):
+        try:
+            self._started_at = time.monotonic()
+            while self.running:
+                time.sleep(self.CHECK_FREQUENCY)
+                self._check_increment_rate()
+                if self._transfer_operation_is_completed():
+                    self.stop()
+                    self._reset()
+        except Exception as exception:  # pylint: disable=broad-except
+            self.log.exception(f"Failed to update transfer rate {self.metric_name!r}")
+            self.metrics.increase("pghoard.transfer_operation.errors")
+
+            self._reset()
+
+    def increment(self, n_bytes: float) -> None:
+        self.metrics.increase(self.metric_name, inc_value=n_bytes, tags=self.tags)
+        self._increments.append(TransferIncrement(n_bytes=n_bytes, tracked_at=time.monotonic()))
+
+    def _check_increment_rate(self) -> None:
+        """
+        Check if the transfer operation is progressing by comparing the time elapsed since
+        last increment with the average time it took for previous increments. If the operation has been inactive,
+        a warning will be logged.
+        """
+        now = time.monotonic()
+
+        avg_rate = 0
+        last_increment_at = self._started_at
+
+        if self._increments:
+            # total "waiting" time between all increments
+            total_increment_diff = sum(
+                next_inc.tracked_at - prev_inc.tracked_at
+                for prev_inc, next_inc in zip(self._increments, self._increments[1:])
+            )
+
+            avg_rate = total_increment_diff / (len(self._increments) - 1) if len(self._increments) > 1 else 0
+            last_increment_at = self._increments[-1].tracked_at
+
+        # log warning in case we have not tracked any progress for the operation since
+        # the last check
+        if last_increment_at and (now - last_increment_at) >= avg_rate + self.WARNING_TIMEOUT:
```

Comment:
**comment** for now I think using the average of "upload" time plus a fixed period of time is reasonable. As it kinda means "Raise a warning if its been more than 5 minutes from what we have expected".  As from what I understand, files are uploaded by chunks... so in case some chunks start experiencing some high latency, that should increase the average and avoid many warnings. But I am open to suggestions if we want to calculate this in a different way :)